### PR TITLE
[WIP] Flux feature parity

### PR DESCRIPTION
**Pull request by MartinuzziFrancesco**
Addressing #12

### CODE REVIEW

File: src/initializers.jl [hunk 1]
```diff
@@ -122,9 +122,191 @@ function truncated_normal(rng::AbstractRNG, ::Type{T}, dims::Integer...; mean=T(
...
+        cols = dims[end]
+    end
+
+    if rows < cols
+        return permutedims(orthogonal(rng, T, cols, rows; gain))
+    end
+
+    mat = randn(rng, T, rows, cols)
+    Q, R = LinearAlgebra.qr(mat)
+    mat .= Array(Q) * sign.(LinearAlgebra.Diagonal(R)) .* T(gain)
```
Comment (avik-pal) [inline src/initializers.jl]: I feel this can be generalized across different devices if we drop the `Array(Q)`

Comment (MartinuzziFrancesco) [reply src/initializers.jl]: good catch, I'll give it a try!

Comment (MartinuzziFrancesco) [reply src/initializers.jl]: it did generalize, nice

**Comment by MartinuzziFrancesco** [comment]
I think I could streamline the tests a little, there are sections that can be shortened a bit. Beside that, `create_bias` is missing, but I think that is more a DL library specific implementation

**Comment by avik-pal** [comment]
yeah `create_bias` is a bit specific. Can you rebase, then we should be good?

**Comment by MartinuzziFrancesco** [comment]
This should be good to go now, I can't see the failed tests on buldkite thou

**Comment by avik-pal** [comment]
I made buildkite public, but rn the cuda clusters have a disk space issue. Let's merge it once that is resolved

File: src/initializers.jl [hunk 2]
```diff
@@ -122,9 +122,185 @@ function truncated_normal(rng::AbstractRNG, ::Type{T}, dims::Integer...; mean=T(
...
+# Arguments
+
+  - `rng::AbstractRNG`: Random number generator.
+  - `T::Type{<:Real}`: The type of the elements in the array.
+  - `dims::Integer...`: The dimensions of the array.
+  - `gain::Number`: Scaling factor for the elements of the orthogonal matrix.
+
+# References
+
+[^Saxe14] Saxe, McClelland, Ganguli. "Exact solutions to the nonlinear dynamics of learning in deep linear neural networks", ICLR 2014, https://arxiv.org/abs/1312.6120
```
Comment (avik-pal) [inline src/initializers.jl]: being a bit nitpicky here but can you all linebreaks manually to make the docstrings match the formatter max linewidth.

Ideally JuliaFormatter should handle this but seems like https://github.com/domluna/JuliaFormatter.jl/issues/434 is not handled.

Comment (MartinuzziFrancesco) [reply src/initializers.jl]: Not a problem, I'll get to it. I really thought that the formatter would handle that

File: src/WeightInitializers.jl [hunk 3]
```diff
@@ -1,6 +1,7 @@
 module WeightInitializers
 
 import PrecompileTools: @recompile_invalidations
+using PartialFunctions, Random, SpecialFunctions, Statistics, LinearAlgebra
```
Comment (avik-pal) [inline src/WeightInitializers.jl]: this line is not needed, right? Just LinearAlgebra needs to be put inside `@recompile_invalidations`

Comment (MartinuzziFrancesco) [reply src/WeightInitializers.jl]: ah yes, I missed a couple of details in the rebase apparently. I fixed that and also added the missing inits to the non differentiable list

**Review by avik-pal — APPROVED** [review]

**Pull request merged**